File: python/questions/trap.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        prem = height[l]
        sufm = height[r]
        ans = 0
        while l < r:
            if prem < sufm:
                ans += prem - height[l]
                l += 1
                prem = max(prem, height[l])
            else:
                ans += sufm - height[r]
                r -= 1
                sufm = max(sufm, height[r])
        return ans;

File: typescript/questions/trap.ts
function trap(height: number[]): number {
  const hl = height.length;
  const st: number[] = [];
  let ans = 0;
  for (let i = 0; i < hl; ++i) {
    const h = height[i];
    while (st.length > 0 && h >= height[st[st.length - 1]]) {
      const bh = height[st.pop()!];
      if (st.length == 0) {
        break;
      }
      ans +=
        (Math.min(h, height[st[st.length - 1]]) - bh) *
        (i - st[st.length - 1] - 1);
    }
    st.push(i);
  }
  return ans;
}
